template/node10-express/function/handler.js

```diff
@@ -1,12 +1,64 @@
 "use strict"
 
-module.exports = (event, context) => {
-    let err;
+// callback version
+const delayCallback = cb => {
+    const min = 1; // 1 sec
+    const max = 5; // 5 sec
+    const delay = Math.round((Math.random() * (max - min) + min));
+    setTimeout(() => cb(delay),  delay * 1000);
+}
+
+module.exports = (event, context, next) => {
```

Comment:
The `next` parameter is new I think? Why is it needed?

Reply:
When we use express we are using a framework, which is in control of the execution of our code, not us. 
Having said that, the `middleware` function at `index.js` is in the "middle" of the execution flow, so we need to "fit" in that schema, node.js coders are used to and aware of that.
I think it's better if we just "build" the response in `handler.js` and let `middleware` in `index.js` take care of sending the response and handling errors. 
Then once you build the response you call `next` to return control to the framework.

Reply:
But let me check again, I think somebody complained about it on Slack

Reply:
The `next` parameter is supposed to be used when you return a callback to the middleware, it's not used when returning a Promise (Promise or Async/Await)

Reply:
I don't line the next() parameter, it's confusing. mixes express and faas programming.
With that said, I believe callbacks should removed altogether, so it doesn't become an issue anymore

Reply:
I know the use of the next() function it is confusing, but is something a express user is used to. I'm open to suggestions on how to solve the callback call in a different way. I already suggested that we need to drop support for it, but until that happens there must be a way to use it.

Reply:
The other way is to pass the res object to the handler function, with passes the control of the flow to it. I think the function main code has to be in control of that.

Reply:
or not use callbacks ;-)

I rebuilt a brand new template - thanks for your inspiration. will publish soon - I am testing right now. It also works differently, more inlined with lambda.


* New better intuitive handler.js
* Super tiny docker image using multistage build, down to 65MB (from 118MB!)
* Garbage collector manually triggered every 30s
* Image properly handles signals using a graceful shutdown
* Catch all uncaught exceptions and unhandled promise rejections

Reply:
Well, that's something I already asked about, what happens if we don't use callbacks at all